feat: add method`from_env` to prometheus exporter builder

LanceEa opened the pull request:
Fixes #242 

Allows configuring the `PrometheusExporter` using environment variables defined in the semantic conventions of the Otel specification. Currently, supports setting the host and port for prometheus in the builder and will fall back to the defaults defined in the otel specification.

LanceEa commented:
This is my first attempt at addressing #242. I'm new to Rust so any suggestions on better approaches are much appreciated. Also, hopefully this aligns with the direction this issues was looking to address.

I have switched it to Draft because I need to get the CLA addressed later tonight when the kids are off to bed :).

TommyCpp commented:
I think the CLA issue is because you didn't link your email in commit to your GitHub account.

TommyCpp commented on opentelemetry-prometheus/src/lib.rs:
```diff
@@ -227,6 +284,8 @@ impl PrometheusExporter {
         default_summary_quantiles: Vec<f64>,
         default_histogram_boundaries: Vec<f64>,
+        host: String,
+        port: String,
```
6 parameters seem a lot in this case. Should we consider deprecate this method in favor of the builder path?

LanceEa replied:
I agree that it felt a little long...the builder pattern feels more consistent with the other exporters.  How does something like this sound?

```rust
#[deprecated(
    since = "v0.9.0",
    note = "Please use the ExporterBuilder to initialize a PrometheusExporter"
)]
```

TommyCpp reviewed (commented):
Looks like a good start, thanks for the effort! 

It may be worth updating the `hyper-prometheus` example to demonstrate users can use `host` and `port` information to start the webserver on the desired port. 

The next step would be to try to solve #293 and use the host/port configuration in the default web server

LanceEa commented on opentelemetry-prometheus/tests/integration_test.rs:
```diff
@@ -159,3 +161,31 @@ fn compare_export(exporter: &PrometheusExporter, mut expected: Vec<&'static str>
     assert_eq!(expected.join("\n"), metrics_only.join("\n"))
 }
+
+#[test]
+fn test_from_env() {
+    let otel_exporter_prometheus_host = "OTEL_EXPORTER_PROMETHEUS_HOST";
+    let otel_exporter_prometheus_port = "OTEL_EXPORTER_PROMETHEUS_PORT";
+
+    // environment variables do not exist
+    env::remove_var(otel_exporter_prometheus_host);
+    env::remove_var(otel_exporter_prometheus_port);
+    let exporter = opentelemetry_prometheus::ExporterBuilder::from_env().init();
+    assert_eq!(exporter.host(), "0.0.0.0");
+    assert_eq!(exporter.port(), "9464");
+
+    // environment variables are available and non-empty strings
+    env::set_var(otel_exporter_prometheus_host, "prometheus-test");
+    env::set_var(otel_exporter_prometheus_port, "9000");
+
+    let exporter = opentelemetry_prometheus::ExporterBuilder::from_env().init();
+    assert_eq!(exporter.host(), "prometheus-test");
+    assert_eq!(exporter.port(), "9000");
+
+    // environment variables are available and empty
+    env::set_var(otel_exporter_prometheus_host, "");
+    env::set_var(otel_exporter_prometheus_port, "");
+    let exporter = opentelemetry_prometheus::ExporterBuilder::from_env().init();
+    assert_eq!(exporter.host(), "0.0.0.0");
+    assert_eq!(exporter.port(), "9464");
+}
```
I had it there at first and then moved it into the integration test to stay consistent. I will move it back.

LanceEa commented on opentelemetry-prometheus/src/lib.rs:
```diff
@@ -122,9 +123,42 @@ pub struct ExporterBuilder {
+    /// Set builder fields for host and port from the OS environment variables
+    ///
+    /// Fields set by developer takes precedent and the OS environment variables will be ignored
+    pub fn from_env() -> Self {
+        let mut builder = ExporterBuilder::default();
+
+        if let Some(host) = env::var("OTEL_EXPORTER_PROMETHEUS_HOST")
+            .ok()
+            .filter(|s| !s.is_empty())
+        {
+            builder = builder.with_host(host)
+        }
+
+        if let Some(port) = env::var("OTEL_EXPORTER_PROMETHEUS_PORT")
+            .ok()
+            .filter(|s| !s.is_empty())
+        {
+            builder = builder.with_port(port);
+        }
+
+        builder
+    }
+
```
Ahh...I didn't see that and was going off from the old comment in the issue. I will go ahead and implement it in the default. 

Also, It looks like the `OTEL_RESOURCE_ATTRIBUTES` was added to the Default implementation of Resource so I don't think I need to do anything special for the prometheus exporter.

LanceEa commented:
Thanks for the review. I will address the feedback tomorrow.

> I think the CLA issue is because you didn't link your email in commit to your GitHub account.

Yes, thanks I need to get me correct email linked.



> It may be worth updating the `hyper-prometheus` example to demonstrate users can use `host` and `port` information to start the webserver on the desired port.

I will take a look at this too and see if I can get it updated.

LanceEa commented:
> It may be worth updating the `hyper-prometheus` example to demonstrate users can use `host` and `port` information to start the webserver on the desired port.

It looks like a standard prometheus pull based scenario where hyper is setup with an endpoint for scraping. This to me aligns with case 2 in the issue where it is a middleware in a web framework.  In that scenario, I don't think these would get used but rather they would be used for either case 1 (push) or case 3 (non-existing web server). Granted since it is part of the public api of `PrometheusExplorer` it could be used but I guess it doesn't feel like the intended use. The spec doesn't provide any specifics about when to use so I didn't see any guidance there. 

So, I think updating or including a new example once #293 lands is probably more appropriate, thoughts?

LanceEa commented:
I have addressed feedback and left comments. I also signed the CLA and got my user-id added to the new commit. It looks like the initial commit is still not associated to my user id. I can squash the commit and clean-up the commit message but I didn't want to confuse the review process. Let me know how you want to proceed!

Cheers,
Lance

TommyCpp commented:
Yeah I think you need to do a force push or something like that. The changes looks good. Will take a closer look on them and the comments later.

TommyCpp commented:
> So, I think updating or including a new example once #293 lands is probably more appropriate, thoughts?

Sounds reasonable to me 👍

LanceEa commented:
Alright, I amended the commit with the missing author. Looks like all the CLA stuff is good to go now. I have updated to rust 1.54, fixed all the new clippy lint issues and rebased so CI should be green now. 

I have left all the individual commits in so that I didn't mess up the PR review history. Does this project squash and cleanup the commit messages before merging into master or would you rather I squash them?

TommyCpp commented on opentelemetry-prometheus/src/lib.rs:
```diff
@@ -122,6 +135,30 @@ pub struct ExporterBuilder {
+    /// The port used by the prometheus exporter
+    ///
+    /// If not set it will be defaulted to port "9464"
+    port: Option<String>,
```
I wonder if it will make sense to use an unsigned int here?

LanceEa replied:
Yeah, I think that makes sense. I did a quick look around to see how the various frameworks and standard library handle it.

`std::net::SocketAddr` treats port as `u16` - https://doc.rust-lang.org/std/net/enum.SocketAddr.html#method.port
`std::net::ToSocketAddr` trait also shows most examples treating it as `u16` - https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html

If I look at a few Web Frameworks they are `u16` becuase they ultimately accept the SocketAddr.
- https://actix.rs/ - leverages a tuple with the port treated like a u16 for conversion to SocketAdd
- https://rocket.rs/v0.5-rc/guide/configuration/#overview
- https://github.com/hyperium/tonic/blob/master/tonic/src/transport/server/mod.rs#L622
- https://docs.rs/hyper/0.14.11/hyper/server/index.html

I can make the switch but I guess the only question is that when parsing the env variable in the impl for `default` how would we handle the potential parse error? I guess it would just get ignored and we fall back to default port? The only problem with that is we are not giving the user any feedback that it occurred so would panic early be a better dev experience?

TommyCpp replied:
That's a good question. I think a good approach is to fall back to default one and print a error message using `global::handle_error` function

LanceEa replied:
Cool, I have gone ahead and made the change and logged an error message.

TommyCpp reviewed (commented):
Awesome work! We do squash and merge on PRs but feel free to squash it now if you want to.

Just one last suggestion and we should be good to go.

LanceEa commented:
I'm not sure what happened with the CLA? I haven't changed anything and I looked and all the commits are associated to the correct email address.

TommyCpp commented:
Oh, I think CLA is down for now. Gonna have to wait for it to come back.

TommyCpp commented:
/easycla

TommyCpp reviewed (approved):
LGTM, good work!
